### ISSUE TITLE
Join room from invitation client-server API fails for Synapse 1.71.0

### ISSUE THREAD

**Issue by manning-ncsa** [matrix-org/synapse#14478]
### Description

I am unable to join a room using an access token on v1.71.0 using the same code that worked on v1.69.0. The Python function I am using is

```python
def accept_invitation(self, room_id=None, access_token=None):
    try:
        assert room_id
        if not access_token:
            access_token = self.conf['auth_token']
        response = requests.post(
            f'''{self.conf['server_url']}/_matrix/client/v3/rooms/{room_id}/join''',
            params = {
                "access_token": access_token,
            }
        )
        return True
    except Exception as e:
        log.error(f'''ERROR accepting invitation: {e}''')
        return False
```

When I provide the access token for the invited Matrix account and the room to be joined, I get a warning in the Synapse server logs as shown below, where no response is ever sent and the account is not allowed to join the room:

```
2022-11-17 19:01:28,329 - synapse.http.servlet - 668 - WARNING - POST-48927- Unable to parse JSON from POST /_matrix/client/v3/join/!roomid:matrix.example.com?access_token=<redacted> response: Expecting value: line 1 column 1 (char 0) (b'')
```

However, a **subsequent identical API call, but using a blank or invalid access token, actually allows the account to join the room**!

```
2022-11-17 19:14:55,712 - synapse.api.auth - 455 - WARNING - POST-49861- Invalid access token in auth: <class 'pymacaroons.exceptions.MacaroonInitException'> Must supply serialized macaroon..

2022-11-17 19:14:55,713 - synapse.http.server - 108 - INFO - POST-49861- <XForwardedForRequest at 0x7f056054aa90 method='POST' uri='/_matrix/client/v3/rooms/!roomid:matrix.example.com/join?access_token=<redacted>' clientproto='HTTP/1.1' site='8008'> SynapseError: 401 - Invalid access token passed.
```

### Steps to reproduce

- Invite an account to a private, unencrypted room.
- Attempt to join the room using the account's access token and the `_matrix/client/v3/rooms/{room_id}/join` API function. The request will stall and the account will not join the room.
- Repeat the request with a blank or invalid access token. The request will fail due to invalid credentials, but the account will now join the room.

### Homeserver

matrix.musesframework.io

### Synapse Version

1.71.0

### Installation Method

Docker (matrixdotorg/synapse)

### Platform

Kubernetes, using [this Helm chart](https://gitlab.com/ananace/charts/-/tree/master/charts/matrix-synapse)

### Relevant log output

```shell
See description above
```

**Comment by DMRobertson** [comment]
# First POST request

> I get a warning in the Synapse server logs as shown below, 
> 
> ```
> 2022-11-17 19:01:28,329 - synapse.http.servlet - 668 - WARNING - POST-48927- Unable to parse JSON from POST /_matrix/client/v3/join/!roomid:matrix.example.com?access_token=<redacted> response: Expecting value: line 1 column 1 (char 0) (b'')
> ```

The warnings comes from https://github.com/matrix-org/synapse/blob/1bf2832714abdfc5e10395e8e76aecc591ad265f/synapse/http/servlet.py#L665-L677

This is caught here and dealt with, but the warning is not suppressed:
https://github.com/matrix-org/synapse/blob/86c5a710d8b4212f8a8a668d7d4a79c0bb371508/synapse/rest/client/room.py#L955-L960

To get rid of the warning, you should send an empty JSON object `{}` as the BODY of your POST request.

> where no response is ever sent and the account is not allowed to join the room:

Your logs aren't sufficient to demonstrate this.

> WARNING - POST-48927- Unable to parse JSON from [...]

Can you please grep for all lines in your Synapse logs matching `POST-48927` and provide them here?

# Second POST request

> 2022-11-17 19:14:55,712 - synapse.api.auth - 455 - WARNING - POST-49861- Invalid access token in auth: <class 'pymacaroons.exceptions.MacaroonInitException'> Must supply serialized macaroon..

Comes from 

https://github.com/matrix-org/synapse/blob/eaed4e6113f5ed40056fa02ae922cb273d02be6e/synapse/api/auth.py#L419-L460

The warning is Synapse's way of saying it doesn't recogise your token.

> However, a subsequent identical API call, but using a blank or invalid access token, actually allows the account to join the room!

Your logs aren't consistent with this.

> 2022-11-17 19:14:55,713 - synapse.http.server - 108 - INFO - POST-49861- <XForwardedForRequest at 0x7f056054aa90 method='POST' uri='/_matrix/client/v3/rooms/!roomid:matrix.example.com/join?access_token=<redacted>' clientproto='HTTP/1.1' site='8008'> SynapseError: 401 - Invalid access token passed.

This suggests we returned HTTP 401 Unauthorized. Can you please provide all logs containing `POST-49861`?

**Comment by clokep** [comment]
As an aside -- it is recommended that you send an access token via headers, not a query parameter. Your above code could be corrected for this (and the empty body) with:

```python
def accept_invitation(self, room_id=None, access_token=None):
    try:
        assert room_id
        if not access_token:
            access_token = self.conf['auth_token']
        response = requests.post(
            f'''{self.conf['server_url']}/_matrix/client/v3/rooms/{room_id}/join''',
            headers = {
                "Authorization": f"Bearer {access_token}",
            },
			json={},
        )
        return True
    except Exception as e:
        log.error(f'''ERROR accepting invitation: {e}''')
        return False
```

**Comment by manning-ncsa** [comment]
Thank you both for your responses. While I still do not understand why the original Python code did not work as it did before, it is possible that a problem with my Redis server that I discovered after creating this issue was actually to blame for the odd behavior. [The improved code](https://github.com/matrix-org/synapse/issues/14478#issuecomment-1319166511) shared by @clokep worked perfectly to join the account to a new room I created to test it. We can close this.
